enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyUtils.java

```diff
@@ -0,0 +1,157 @@
+        // Base 64 decode key data.
+        String decodedKeyData = new String(Base64.getDecoder().decode(keyData));
+        // Convert data into JSON.
+        JSONObject jsonObject = (JSONObject) JSONValue.parse(decodedKeyData);
+        // Extracting the key.
+        String key = jsonObject.getAsString(APIKeyConstants.API_KEY_JSON_KEY);
+        // Return SHA256 hash of the key.
+        return DigestUtils.sha256Hex(key);
+    }
+
+    /**
+     * This function exchanges a given API key to an JWT token.
+     *
+     * @param keyHash    Key hash
+     * @return          JWT corresponding to given key hash.
+     */
+    public static Optional<String> exchangeAPIKeyToJWT(String keyHash) {
+
+        URL url = null;
+        try {
+            url = new URL(ConfigHolder.getInstance().getConfig().getApiKeyConfig().getOauthAgentURL());
+        } catch (MalformedURLException e) {
+            log.error("Error occurred while parsing OAuth agent URL", e);
+            return Optional.empty();
+        }
+        try (CloseableHttpClient httpClient = (CloseableHttpClient) FilterUtils.getHttpClient(url.getProtocol())) {
+            // Create a request to exchange API key to JWT.
+            HttpPost exchangeRequest = new HttpPost(url.toURI());
+            exchangeRequest.addHeader("Content-Type", ContentType.APPLICATION_JSON.toString());
+            exchangeRequest.setEntity(new StringEntity("{\"apiKeyHash\": \"" + keyHash + "\"}"));
+            try (CloseableHttpResponse response = httpClient.execute(exchangeRequest)) {
+                if (response.getStatusLine().getStatusCode() == 200) {
+                    HttpEntity entity = response.getEntity();
+                    try (InputStream content = entity.getContent()) {
+                        OAuthAgentResponse resp = new Gson().fromJson(IOUtils.toString(content),
+                                OAuthAgentResponse.class);
+                        return Optional.of(resp.getAccessToken());
+                    }
+                }
+            }
+        } catch (Exception e) {
+            log.error("Error occurred while exchanging API key to JWT", e);
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Validate JWT expiry.
+     *
+     * @param jwt   JWT
+     * @return      True if expired.
+     */
+    public static boolean isJWTExpired(String jwt) {
+
+        try {
+            SignedJWT signedJWT = SignedJWT.parse(jwt);
+            Date now = new Date();
+            Date exp = signedJWT.getJWTClaimsSet().getExpirationTime();
+            return DateUtils.isBefore(exp, now, ConfigHolder.getInstance().getConfig()
+                    .getPatConfig().getTokenExpirySkewSeconds());
+        } catch (ParseException e) {
+            log.debug("JWT token parsing failed");
+        }
+        return true;
+    }
+
+    private static String generateCRC32Checksum(String data) {
+
+        CRC32 crc32 = new CRC32();
```

Comment:
can't this a a class level static variable ?

Reply:
Since this object itself generates the checksum with the `update()` function and `getValue()` will return the generated checksum for the given bytes array. In order to reuse the same object for new checksum generations, we could use `reset()` and `update()`. But this could lead to concurrency issues when same object is being used by two concurrent requests.